crossplane-contrib/provider-kubernetes
# Add Alpha Support for Server Side Apply

### Description of your changes

This PR introduces a new flag, `--enable-server-side-apply`, to enable [Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) instead of the existing Patching Applicator combined with the last applied configuration annotation. 

Fixes #37
Fixes #114 

I have:


### How has this code been tested

Tested/verified with `examples/object/object.yaml` and `examples/object/object-service.yaml` with/without the flag.

Verified that enabling the feature does not set last-applied configuration and properly sets field manager in the managed fields:

1. Create the `Object` example wrapping a k8s `Service`:
```bash
❯ kubectl create -f examples/object/object-service.yaml
object.kubernetes.crossplane.io/sample-service created
```
2. Check the created service with managed fields:
```bash
❯ kubectl get svc  sample-service -o yaml --show-managed-fields
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: "2023-08-23T10:29:35Z"
  labels:
    example: "true"
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:example: {}
      f:spec:
        f:ports:
          k:{"port":80,"protocol":"TCP"}:
            .: {}
            f:port: {}
            f:protocol: {}
            f:targetPort: {}
        f:selector: {}
    manager: crossplane-provider-kubernetes
    operation: Apply
    time: "2023-08-23T10:29:35Z"
  name: sample-service
  namespace: default
  resourceVersion: "32113"
  uid: f540e8ed-4ccc-4fc2-a38f-bdb76658bc96
spec:
  clusterIP: 10.96.125.58
  clusterIPs:
  - 10.96.125.58
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - port: 80
    protocol: TCP
    targetPort: 9376
  selector:
    app.kubernetes.io/name: MyApp
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```
3. Make an update on the `Object`.
4. Verify the update reflected on the `Service`.

[contribution process]: https://git.io/fj2m9
